// @vinejs/vine/build/src/schema/union/main.js
import camelcase from 'camelcase';
import { messages } from '../../defaults.js';
import { OTYPE, COTYPE, PARSE } from '../../symbols.js';
export class VineUnion {
    #conditionals;
    #otherwiseCallback = (_, field) => {
        field.report(messages.union, 'union', field);
    };
    constructor(conditionals) {
        this.#conditionals = conditionals;
    }
    otherwise(callback) {
        this.#otherwiseCallback = callback;
        return this;
    }
    clone() {
        const cloned = new VineUnion(this.#conditionals);
        cloned.otherwise(this.#otherwiseCallback);
        return cloned;
    }
    [PARSE](propertyName, refs, options) {
        return {
            type: 'union',
            fieldName: propertyName,
            propertyName: options.toCamelCase ? camelcase(propertyName) : propertyName,
            elseConditionalFnRefId: refs.trackConditional(this.#otherwiseCallback),
            conditions: this.#conditionals.map((conditional) => conditional[PARSE](propertyName, refs, options)),
        };
    }
}
