// @vinejs/compiler/build/src/compiler/nodes/tuple.js
import { BaseNode } from './base.js';
import { defineArrayGuard } from '../../scripts/array/guard.js';
import { defineIsValidGuard } from '../../scripts/field/is_valid_guard.js';
import { defineFieldNullOutput } from '../../scripts/field/null_output.js';
import { defineFieldValidations } from '../../scripts/field/validations.js';
import { defineArrayInitialOutput } from '../../scripts/array/initial_output.js';
import { defineFieldExistenceValidations } from '../../scripts/field/existence_validations.js';
export class TupleNodeCompiler extends BaseNode {
    #node;
    #buffer;
    #compiler;
    constructor(node, buffer, compiler, parent, parentField) {
        super(node, compiler, parent, parentField);
        this.#node = node;
        this.#buffer = buffer;
        this.#compiler = compiler;
    }
    #compileTupleChildren() {
        const buffer = this.#buffer.child();
        const parent = {
            type: 'tuple',
            fieldPathExpression: this.field.fieldPathExpression,
            outputExpression: this.field.outputExpression,
            variableName: this.field.variableName,
            wildCardPath: this.field.wildCardPath,
        };
        this.#node.properties.forEach((child) => {
            this.#compiler.compileNode(child, buffer, parent);
        });
        return buffer.toString();
    }
    compile() {
        this.defineField(this.#buffer);
        this.#buffer.writeStatement(defineFieldExistenceValidations({
            allowNull: this.#node.allowNull,
            isOptional: this.#node.isOptional,
            variableName: this.field.variableName,
        }));
        const isArrayValidBlock = defineIsValidGuard({
            variableName: this.field.variableName,
            bail: this.#node.bail,
            guardedCodeSnippet: `${defineArrayInitialOutput({
                variableName: this.field.variableName,
                outputExpression: this.field.outputExpression,
                outputValueExpression: this.#node.allowUnknownProperties
                    ? `copyProperties(${this.field.variableName}.value)`
                    : `[]`,
            })}${this.#compileTupleChildren()}`,
        });
        const isValueAnArrayBlock = defineArrayGuard({
            variableName: this.field.variableName,
            guardedCodeSnippet: `${defineFieldValidations({
                variableName: this.field.variableName,
                validations: this.#node.validations,
                bail: this.#node.bail,
                dropMissingCheck: true,
            })}${this.#buffer.newLine}${isArrayValidBlock}`,
        });
        this.#buffer.writeStatement(`${isValueAnArrayBlock}${this.#buffer.newLine}${defineFieldNullOutput({
            allowNull: this.#node.allowNull,
            outputExpression: this.field.outputExpression,
            variableName: this.field.variableName,
            conditional: 'else if',
        })}`);
    }
}
